Allow toJSON types to be used with atomFamily and selectorFamily

BenjaBobs opened the pull request:
This allows such use cases as using Date as params for a family, e.g. if you want to create a time slice of a range of data.
This fixes #760

drarmstr reviewed (commented):
I'm used to Flow now, assuming TypeScript uses inexact types by default?  I can add similar typing (though inexact object) to the Flow types.  Thank you for adding the static TypeScript test.  Would you be up for adding actual Jest unit tests in `Recoil_selectorFamily-test.js` and `Recoil_atomFamliy-test.js` for using `Date`?

BenjaBobs commented:
Yes, it just means anything that has a `toJSON` property that is a `function` that returns a `string` is valid.

Sure, I'll look into it. 👍

BenjaBobs commented:
@drarmstr I added the tests you suggested, and rebased onto master. As for the code style of the tests, I tried as best I could to follow the precedent set by you guys, but will be happy to fix any issues/discrepancies you may find.

drarmstr commented:
> @drarmstr I added the tests you suggested, and rebased onto master. As for the code style of the tests, I tried as best I could to follow the precedent set by you guys, but will be happy to fix any issues/discrepancies you may find.

Wonderful!  Importing now for landing.
